### PR TITLE
refactor(parsing): consolidate the fence state machine into lines.ts; memory sections via the shared parser

### PR DESCRIPTION
**Pull request by aliasunder**
## What

Dedup follow-up to #168 — the second half of the parser-layer restructure. Consolidates the duplicated low-level parsing primitives into one home, with **two Obsidian-parity fixes** as the only behavior changes (both strict upside).

### New `obsidian-markdown/lines.ts`
The single home for low-level line/fence logic:
- `advanceFence` — the one CommonMark §4.5 fenced-code state machine.
- `splitIntoLines` — CRLF-safe line splitting.
- `classifyLines` — the `{ text, inCode }` generator, **moved out of `links.ts`** (it had zero link logic; it lived there only because link rewriting needed it first).

### Three fence state machines → one
`links.ts` and `headings.ts` (which had **two** private copies — inside `parseHeadings` and `findTrailingCommentBlockStart`) now thread fence state through `advanceFence`. The long-standing docstring claim that this is "the single home of the fence state machine" is finally true.

### `memory-store.parseSections` → shared `parseHeadings`
Rebuilt on the shared heading parser (filtered to H1/H2) with a separate entry-count pass, replacing a hand-rolled, near-duplicate section parser.

### Parity fixes (the only behavior changes)
- **Indented fences hide headings.** `headings.ts` previously matched fences only at column 0; the shared grammar allows the CommonMark 0–3 space indent, so a `# heading`-looking line inside an indented code fence is no longer mis-parsed as a heading.
- **Memory sections are fence-aware.** A `## ...`-looking line inside a code fence in an `About Me/` file is no longer mistaken for a real section.

### CRLF unification
`splitIntoLines` replaces hand-rolled `"\n"` splits at the note-body production sites (`vault-filesystem`, `search-index`, `vault-patcher`, `memory-store`), so CRLF-authored notes stop leaking a stray `\r` into heading/section parsing.

## Tests

- New `lines.test.ts` covering `splitIntoLines`, `advanceFence`, and `classifyLines` (the latter relocated from `links.test.ts`).
- Two-bar parity tests for both fixes — each fails under the old code for the right reason.

## Verification

- `npm run build` (tsc, server + cli) ✅
- `npm test` — **915 passed** ✅
- `npm run lint` ✅

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Fixed incorrect detection of headings inside fenced code blocks; headings are now properly ignored when within code fences.
  * Improved handling of Windows-style line endings (CRLF) across file parsing.

* **Tests**
  * Added comprehensive test coverage for Markdown parsing edge cases including indented code blocks and fence delimiter handling.

* **Refactor**
  * Consolidated shared line and fence-parsing logic to reduce duplication across modules.

### CODE REVIEW

File: src/vault-mcp/obsidian-markdown/__tests__/headings.test.ts [hunk 1]
```diff
@@ -48,6 +48,23 @@ describe("parseHeadings", () => {
...
+    // grammar. The previous heading-local matcher required column 0, so it would
+    // have mis-parsed "# Not a heading" here as a real heading.
+    const lines = [
+      "# Real",
+      "   ```",
+      "# Not a heading",
+      "   ```",
+      "## Also real",
+    ]
+    expect(parseHeadings(lines).map((h) => h.text)).toEqual([
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/obsidian-markdown/__tests__/headings.test.ts]: Fixed in d65255c — renamed the new tests' callbacks to `heading` / `outline` per the explicit-naming convention. The pre-existing `h`/`o` in older tests were left untouched to keep the change in scope.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/vault-mcp/vault-operations/memory-store.ts [hunk 2]
```diff
@@ -123,61 +125,28 @@ type ParsedSection = Readonly<{
...
+    .map((heading) => ({
+      heading: heading.text,
+      level: heading.level as 1 | 2,
+      startLine: heading.startLine,
+      bodyStartLine: heading.bodyStartLine,
+      bodyEndLine: heading.bodyEndLine,
+      entryCount: lines
+        .slice(heading.bodyStartLine, heading.bodyEndLine)
+        .filter((bodyLine) => ENTRY_PATTERN.test(bodyLine)).length,
+    }))
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/vault-operations/memory-store.ts:166]: Fixed in d65255c — `parseSections` now computes `entryCount` only for H2 sections (the H1's is discarded by `listMemoryFiles`), and via a single-pass loop instead of `slice(...).filter(...)`, so the H1 title section's whole-file scan and the per-section allocations are gone.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by aliasunder** [comment]
/@coderabbit
